File: 2021_04_18_make_a_lisp_in_csharp/Mal/Mal/Env.cs

using PowerFP;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using static Mal.Printer;
using static Mal.Types;

namespace Mal
{
    public static class EnvM
    {
        public record Env
        {
            public Map<Symbol, MalType> Data { get; internal set; }
            public Env? Outer { get; init; }

            public Env(Map<Symbol, MalType> data, Env? outer)
            {
                Data = data;
                Outer = outer;
            }
        }

        public static MalType Set(this Env env, Symbol key, MalType value)
        {
            env.Data = env.Data.Add(key, value);
            return value;
        }


        public static Env? Find(this Env env, Symbol key)
            => FindEnvAndValue(env, key) switch
            {
                null => null,
                (var Env, _) => Env
            };

        public static MalType Get(this Env env, Symbol key)
            => FindEnvAndValue(env, key) switch
            {
                null => throw new Exception($"Cannot find symbol '{key}' in Env"),
                (_, var Value) => Value
            };

        private static (Env, MalType)? FindEnvAndValue(this Env env, Symbol key)
            => env.Data.TryFind(key) switch
            {
                (true, var value) => (env, value!),
                _ => env.Outer == null ? null : FindEnvAndValue(env.Outer, key)
            };
    }
}

File: 2021_10_21_make_a_lisp_in_fsharp/MiniMal/MiniMal/Env.fs
module Env

open Types

type Env(data: Map<string, MalType>, outer: Env option) =
    let mutable data = data
    let outer = outer

    let rec findEnvAndValue (env: Env) key =
        match Map.tryFind key env.Data with
        | Some (value) -> Some(env, value)
        | None ->
            env.Outer
            |> Option.bind (fun outerEnv -> findEnvAndValue outerEnv key)

    member private this.Data = data
    member private this.Outer = outer

    member this.Set key value =
        data <- Map.add key value data
        value

    member this.Find key =
        findEnvAndValue this key
        |> Option.map (fun (foundEnv, _) -> foundEnv)

    member this.Get key =
        match findEnvAndValue this key with
        | Some (_, foundMal) -> foundMal
        | None -> failwith $"Cannot find symbol '{key}' in Env"
